models/validationAndNormalization.go

```diff
@@ -0,0 +1,431 @@
+    err := runUserValFns(user,
+        userValid.passwordMinLength,
+        userValid.bcryptPassword,
+        userValid.passwordHashRequired,
+        userValid.rememberMinBytes,
+        userValid.hmacRemember,
+        userValid.rememberHashRequired,
+        userValid.normalizeEmail,
+        userValid.requireEmail,
+        userValid.emailFormat,
+        userValid.emailIsAvail,
+    )
+    if err != nil {
+        return err
+    }
+
+    return userValid.UserDB.Update(user)
+}
+
+/**
+ * @brief:  Delete provided user with ID
+ *
+ * @param:  id - User to be deleted
+ *
+ * @return: nil on success
+ *          ErrIDInvalid if ID is invalid
+ *          Else error
+ **/
+func (userValid *userValidator) Delete(id uint) error {
+    var user User
+    user.ID = id
+
+    err := runUserValFns(&user,
+        userValid.validId,
+    )
+    if err != nil {
+        return err
+    }
+
+    return userValid.UserDB.Delete(id)
+}
+
+/**
+ * @brief:  Hash the remember token
+ *
+ * @param:  token - Token of user
+ *
+ * @return: User on success, else error
+ **/
+func (userValid *userValidator) ByRemember(token string) (*User, error) {
+    user := User {
+        Remember: token,
+    }
+
+    err := runUserValFns(&user,
+        userValid.hmacRemember,
+    )
+    if err != nil {
+        return nil, err
+    }
+
+    return userValid.UserDB.ByRemember(user.RememberHash)
+}
+
+/**
+ * @brief:  Normalize email address
+ *
+ * @param:  email - Email address that is going to be normalized
+ *
+ * @return: User on success, else error
+ **/
+func (userValid *userValidator) ByEmail(email string) (*User, error) {
+    user := User {
+        Email:  email,
+    }
+
+    err := runUserValFns(&user, userValid.normalizeEmail)
+    if err != nil {
+        return nil, err
+    }
+
+    return userValid.UserDB.ByEmail(user.Email)
+}
+
+/* ==============================*/
+/*      METHODS TO VALIDATION    */
+/*        AND NORMALIZATION      */
+/* ==============================*/
+
+/**
+ * @brief:  Checks to see if password is provided
+ *
+ * @param:  user - Contains password
+ *
+ * @return: nil on success, else ErrPasswordRequired
+ **/
+func (userValid *userValidator) passwordRequired(user *User) error {
+    if user.Password == "" {
+        return ErrPasswordRequired
+    }
+
+    return nil
+}
+
+/**
+ * @brief:  Checks to see if users password hash has a value
+ *
+ * @param:  user - Contains password hash
+ *
+ * @return: nil on success, else error
+ **/
+func (userValid *userValidator) passwordHashRequired(user *User) error {
+    if user.PasswordHash == "" {
+        return ErrPasswordRequired
+    }
+
+    return nil
+}
+
+/**
+ * @brief:  Check if the password is greaterc than min length
+ *
+ * @param:  user - Contains textbased password
+ *
+ * @return nil on success, else error
+ **/
+func (userValid *userValidator) passwordMinLength(user *User) error {
+    if user.Password == "" {
+        return nil
```

Comment:
It would be better to return `ErrPasswordRequired` instead of `nil` which shows success when an empty password isn't greater than min length

Reply:
Added error message